[WIP] adding errorCallback to native modules for errors throw on status-go side

siddarthkay opened the pull request:
fixes #14720

### Review notes
This is probably the only way to handle an error call back.
Status-go returns everything as 1 string.. I would not want to change that since Desktop team also relies on status-go.

#### Platforms
- Android
- iOS

#### Areas that maybe impacted
Any and everything where the app uses status-go

### Steps to test
Please do a full test

status: ready

siddarthkay committed:
adding errorCallback to deserializeAndCompressKey

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | adc5318f | [#1](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-14722/1/) | 2023-01-06 07:48:11 | ~2 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-230106-074545-adc531-pr14722-tests.log) |

siddarthkay committed:
revert podfile.lock

siddarthkay committed:
trying to make ErrorCallback optional

siddarthkay committed:
add error-callback to most native module usages

siddarthkay committed:
fixed a few danglers

siddarthkay committed:
add errorCallbacks on Objective C side

siddarthkay committed:
lint-fix and pass tests

cammellos commented on modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:
```diff
@@ -531,33 +531,37 @@ public void run() {
         Runnable runnableTask = () -> {
             String res = method.get();
-            callback.invoke(res);
+             if (res.contains("error")) {
```
This is very dangerous, what if you receive a message with `error` in the content? we need to parse it as JSON (could be optimized by scanning the string, but don't think it's worth the risk)

siddarthkay replied:
hmmm, every method in status-go uses `makeJSONResponse ` method 
which will add the error key to the response string :  reference https://github.com/status-im/status-go/blob/ec7c0e9c7d34f73d89d8fb4be0e116aa29c3a29d/mobile/status.go#L625

siddarthkay replied:
> This is very dangerous, what if you receive a message with `error` in the content? we need to parse it as JSON (could be optimized by scanning the string, but don't think it's worth the risk)

Hmmm I see your point

siddarthkay replied:
I've done this on Android native module, I will work on the IOS side as well! Thanks for the suggestion @cammellos

siddarthkay replied:
> This is very dangerous, what if you receive a message with `error` in the content? we need to parse it as JSON (could be optimized by scanning the string, but don't think it's worth the risk)

For ios side I converted the response to a dictionary and used a similar approach here, this should ensure we look for the error key and we do not trigger false positives by "error" in message

cammellos commented on src/status_im/browser/core.cljs:
```diff
@@ -534,7 +537,10 @@
+        (callback nil (.parse js/JSON response))))
+    (fn [error-message]
+      (log/debug "error while status/call-rpc" error-message))
+   )))
```
linting

siddarthkay committed:
get rid of dangling parens again

cammellos commented on src/status_im/keycard/real_keycard.cljs:
```diff
@@ -321,7 +321,10 @@
      (let [{:keys [error]} (types/json->clj result)]
        (if-not (string/blank? error)
```
if we start using the callbacks, we should not handle errors though on the on-success, or that's still an eventuality? Similarly in other parts of the code

siddarthkay replied:
yes this will definitely need some more refactoring :) 
But since i did not write this functionality , I was hesitant on modifying it.. I will look at this one more time :)

cammellos reviewed (changes requested):
I think we need to be more careful in detecting an error, this will give false positives

siddarthkay committed:
use jsonobject to check error key

siddarthkay committed:
convert response to dictionary to find error key

This is done by a custom and handy helper method i wrote :  `(BOOL)responseContainsErrorKey`

siddarthkay committed:
get rid of error handling in on-success callbacks

siddarthkay committed:
fix make lint warnings

ilmotta commented on src/status_im/keycard/simulated_keycard.cljs:
```diff
@@ -402,28 +407,29 @@
+                   (fn [error-message]
+                     (on-failure error-message))))))
+            (fn [error-message]
+              (log/debug "error while status/multiaccount-derive-addresses" error-message)))))
```
@siddarthkay In this PR I see a lot of `log/debug` in *error callback* functions. They seem to swallow the error, because they don't rethrow it. Are the errors really propagating up the call tree? If not, then I think this is a more serious problem.

If they propagate correctly, then it shouldn't be necessary to add intermediary calls to `log/debug`, only a top level error handler would be enough. In Java, this would be equivalent of doing a try/catch in intermediary methods that don't actually do anything with the error (except logging). This is a known anti-pattern.

It's probably out of scope for your PR, but these nested callbacks for success/error could be dramatically improved in the future if we used the de facto standard in Javascript (promises), or some other means in Clojure land. This would give us a sane way to handle errors and create composable error handlers.

siddarthkay replied:
> dramatically improved in the future if we used the de facto standard in Javascript (promises)

I did look into the native modules side on React Native and it does support Promises and would make the code a lot simpler but would require a lot of re-write ( same as this PR 🫣 ).

I would for now log a new issue to address this concern later.

siddarthkay replied:
> Are the errors really propagating up the call tree? If not, then I think this is a more serious problem.

yes they do and that was the inspiration to introduce error-callbacks to avoid handling of both success and error cases in one callback.

ilmotta replied:
> I would for now log a new issue to address this concern later.

Thanks, better in a separate issue as you said.

qfrank commented on modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:
```diff
@@ -531,33 +531,43 @@ public void run() {
+    private void executeRunnableStatusGoMethod(Supplier<String> method, Callback successCallback,Callback errorCallback) throws JSONException {
         if (!checkAvailability()) {
-            callback.invoke(false);
+            successCallback.invoke(false);
```
it makes me a little confused here, if we can't execute `method`, why not use `errorCallback` instead but `successCallback`?

siddarthkay replied:
> it makes me a little confused here, if we can't execute `method`, why not use `errorCallback` instead but `successCallback`?

Well I need to do some more research on this but what this basically means is that when we only had a single callback we used to do the callback anyways with a false return whenver `checkAvailability` would return `false` .
But I could definitely make this better.

thanks for pointing it out @qfrank  :)

qfrank commented on src/status_im/browser/core.cljs:
```diff
@@ -418,7 +418,10 @@
-                                                        :result    (types/json->clj %)}])}}
+                                                        :result    (types/json->clj %)}])
+                                      :on-error     (fn [error-message]
+                                                      (log/debug "error while status/recover-message"
```
if it's an error, i prefer to use log/warn or log/error rather than log/debug

qfrank commented on src/status_im/multiaccounts/create/core.cljs:
```diff
@@ -40,7 +41,7 @@
 (re-frame/reg-fx
  ::store-multiaccount
- (fn [[id key-uid hashed-password callback]]
+ (fn [[id key-uid hashed-password callback on-error]]
```
can we rename callback to on-success so that we can c the `good pair` on-success/on-error? 😅

qfrank commented on src/status_im/signing/core.cljs:
```diff
@@ -25,10 +25,11 @@
    (status/send-transaction (types/clj->json tx-obj)
                             hashed-password
-                            cb)))
+                            cb
```
can we also rename `cb` here ?

siddarthkay replied:
> can we also rename `cb` here ?

Well it was always `cb` from the beginning, but sure why not, I'll improve this :)

ilmotta replied:
On the Clojure side, more specifically re-frame events, it's preferred to use more explicit names, like `on-error`, `on-success`, particularly because re-frame uses this convention throughout its documentation. There are also external effects, like (re-frame-http-fx) who follow this convention (although they use `on-failure` instead of `on-error`). Additionally, pretty much all effects in Status codebase use this naming convention (`:rpc/call` being the most famous).

I think we should keep the consistency and use `on-error` and `on-success` as the names of such arguments.

siddarthkay commented:
Started working on this again

siddarthkay committed:
Revert "fix make lint warnings"

This reverts commit a7580dbab3c098f2e3695caebc23ee96fa16c4d8.

siddarthkay committed:
Revert "get rid of error handling in on-success callbacks"

This reverts commit 7740119e871f29dcbbd6a29636076d7e0681be2e.

siddarthkay committed:
refactor verify method

get rid of error handling in on-success callbacks for this function

fix event add-new-account-password-verifyied

fix up this ambitious change

fix lint and closing parenthesis

flexsurfer commented:
hey @siddarthkay still relevant?

siddarthkay commented:
@flexsurfer : yes it has some work in progress which i plan to merge this week

siddarthkay commented:
I'll convert to Draft

flexsurfer commented:
feel free to reopen when work in progress again
